pennylane_lightning/src/simulator/Measures.hpp

```diff
@@ -0,0 +1,218 @@
+// Copyright 2021 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @file
+ * Defines a class for the measurement of observables
+ * in quantum states represented by the StateVector class.
+ */
+
+#pragma once
+
+#include <complex>
+#include <stdio.h>
+#include <vector>
+
+#include "StateVector.hpp"
+#include "Util.hpp"
+
+/// @cond DEV
+namespace {
+using std::complex;
+using std::size_t;
+using std::vector;
+}; // namespace
+/// @endcond
+
+namespace Pennylane {
+using namespace Util;
+/**
+ * @brief Observable's Measurement Class.
+ *
+ * This class performs measurements in the state vector provided to its
+ * constructor. Observables are defined by its operator(matrix) or through a
+ * string-based function dispatch.
+ *
+ * @tparam fp_t Floating point precision of underlying measurements.
+ */
+template <class fp_t = double> class Measures {
+  private:
+    StateVector<fp_t> measured_statevector;
+
+  public:
+    Measures(StateVector<fp_t> provided_statevector)
+        : measured_statevector{provided_statevector} {};
+
+    /**
+     * @brief Probabilities of each computational basis state.
+     *
+     * @return Floating point std::vector with probabilities
+     * in lexicographic order according to wires.
+     */
+    vector<fp_t> probs() {
+        complex<fp_t> *arr_data = measured_statevector.getData();
+        vector<double> basis_probs(measured_statevector.getLength(), 0);
+        for (size_t ind = 0; ind < 8; ind++) {
+            basis_probs[ind] = std::pow(std::abs(arr_data[ind]), 2.0);
+        }
```

Comment:
The 8 here is wrong, I will fix that.

Reply:
The `basis_prob` can be a member, or maybe calculated on-the-fly for the Probabilities for a subset of the full system.

Reply:
```
std::transform(arr_data, arr_data+measured_statevector.getLength(), basis_probs.begin(),
                   [](const std::complex<fp_t> &z) -> fp_t { return std::norm(z); } );
```
Beautiful suggestion. Done!

Reply:
Regarding making the basis_prob a class member, I gave some think on this and I don't think it is safe to be done. The StateVector we are measuring is not affect by measure by now, but this can change in the future. If we store the basis_probability for future reference, when we come to use it, the data can be out-dated. I mean, it is easy to be implemented, though.